### PR TITLE
FP16 Support for Conv Act and FullyConnected Layers

### PR DESCRIPTION
**Pull request by piiswrong**

### CODE REVIEW

**Comment by tqchen** [comment]
please check the travis, and add a few test cases to confirm that it works

**Comment by piiswrong** [comment]
@tqchen Let's do a final review and merge

File: include/mxnet/operator.h [hunk 1]
```diff
@@ -287,6 +287,17 @@ class OperatorProperty {
    */
   virtual Operator* CreateOperator(Context ctx) const = 0;
   /*!
+   * \brief Create a Operator on specific context and input type
+   */
```
Comment (tqchen) [inline include/mxnet/operator.h]: document the parameters, what are in_shape and in_type?

File: python/mxnet/symbol.py [hunk 2]
```diff
@@ -633,7 +633,7 @@ def simple_bind(self, ctx, grad_req='write', type_dict=None, **kwargs):
         if grad_req != 'null':
             grad_ndarrays = {}
             for name, shape, dtype in zip(self.list_arguments(), arg_shapes, arg_types):
-                if not (name.endswith('data') or name.endswith('label')):
+                if grad_req is not dict or grad_req[name] != 'null':
```
Comment (tqchen) [inline python/mxnet/symbol.py]: What does is not dict mean here?

Comment (tqchen) [inline python/mxnet/symbol.py]: use ```not isinstance(grad_req, dict)```

File: include/mxnet/operator.h (hunk 1 above)
Comment (tqchen) [inline include/mxnet/operator.h]: Maybe document that eventually we will deprecate the CreateOperator function and move to CreateOpEx

**Comment by tqchen** [comment]
The Logic LGTM.  Please do the following things

- Fix my comments
- Explain now how type information is specified in the simple_bind, I am a bit confused in the API, since both shape and type need to be specified.

I want to make sure the people are happy with the new typed simple_bind API, so please provide a an example of bind and simple_bind, so we understand what is going on and can know what to expect.

Also cc @pluskid @antinucleon @mavenlin for review he typed API.

Thanks for the great effort

**Comment by futurely** [comment]
How to use FP16 to train a new model or further fine tune an existing model?

**Comment by futurely** [comment]
After updating to the latest version, compilation failed with the following errors.
https://gist.github.com/futurely/eed33c95d8c334b2c878

**Comment by futurely** [comment]
The commit before this PR https://github.com/dmlc/mxnet/commit/5224f6fad6680a6fe0824fea58a822b9bc284fef can build.

**Pull request merged**